README.md
# chronos-python

[![Build Status](https://travis-ci.org/asher/chronos-python.svg?branch=master)](https://travis-ci.org/asher/chronos-python)

This is a Python client library for the [Chronos](https://mesos.github.io/chronos/docs/api.html) HTTP [Rest API](https://mesos.github.io/chronos/docs/api.html)

## Installation

    pip install chronos-python
    # or
    git clone git@github.com/asher/chronos-python
    python setup.py install

## Usage Examples


Create a ``ChronosClient``

    >>> import chronos
    >>> client = chronos.connect("chronos.mesos.server.com:8080")
    # or specify multilple servers that will be tried in order
    >>> client = chronos.connect(["chronos1.mesos.server.com:8080", "chronos2.mesos.server.com:8080"])

List all jobs:

     >>> client.list()
     [{u'softError': False, u'scheduleTimeZone': u'null', u'successCount': 702, u'cpus': 0.25, u'disabled': False, u'ownerName': u'', u'owner': u'noop', u'disk': 256.0, u'errorCount': 0, u'container': {u'image': u'my-docker-registry:443/myimage', u'type': u'docker', u'network': u'BRIDGE', u'volumes': []}, u'errorsSinceLastSuccess': 0, u'highPriority': False, u'dataProcessingJobType': False, u'arguments': [], u'uris': [u'file:///root/.dockercfg'], u'shell': True, u'description': u'', u'schedule': u'R/2015-12-18T10:40:00.000Z/PT10M', u'mem': 1024.0, u'epsilon': u'PT60S', u'retries': 2, u'name': u'my job 1', u'runAsUser': u'root', u'lastSuccess': u'2015-12-18T10:30:09.755Z', u'environmentVariables': [], u'executorFlags': u'', u'command': u'sensu-scheduled-canary.sh', u'executor': u'', u'async': False, u'lastError': u'', u'constraints': []}, {u'softError': False, u'scheduleTimeZone': u'null', u'successCount': 40, u'cpus': 0.25, u'disabled': False, u'ownerName': u'', u'owner': u'noop', u'disk': 256.0, u'errorCount': 0, u'container': {u'image': u'my-docker-regsitry:443/myimage', u'type': u'docker', u'network': u'BRIDGE', u'volumes': [], u'errorsSinceLastSuccess': 0, u'highPriority': False, u'dataProcessingJobType': False, u'arguments': [], u'uris': [u'file:///root/.dockercfg'], u'shell': True, u'description': u'', u'schedule': u'R/2015-12-18T11:00:00.000Z/PT60M', u'mem': 1024.0, u'epsilon': u'PT60S', u'retries': 2, u'name': u'example_service mesosstage_kwabatch gitfb0c7ac5 config95bc9b2f', u'runAsUser': u'root', u'lastSuccess': u'2015-12-18T08:00:12.965Z', u'environmentVariables': [], u'executorFlags': u'', u'command': u'echo "This batch should run once per hour, and take 2 hours" && sleep 2h', u'executor': u'', u'async': False, u'lastError': u'', u'constraints': []}]


Add a new job:

    >>> job = { 'async': False, 'command': 'echo 1', 'epsilon': 'PT15M', 'name': 'foo',
    'owner': 'me@foo.com', 'disabled': True, 'schedule': 'R/2014-01-01T00:00:00Z/PT60M'}
    >>> client.add(job)

Update an existing job:

    >>> job = { 'async': False, 'command': 'echo 1', 'epsilon': 'PT15M', 'name': 'foo',
    'owner': 'me@foo.com', 'disabled': True, 'schedule': 'R/2014-01-01T00:00:00Z/PT60M'}
    >>> client.update(job)

Run a job:

    >>> client.run("job123")

Delete a job:

    >>> client.delete("job123")

Delete all the in flight tasks for a job:


    >>> client.delete_tasks("job123")


## Included Scripts
* `chronos-sync-jobs.py` - Sync chronos jobs from a directory tree containing job.json files.
`chronos-sync-jobs.py --hostname chronos.server.com:4400 --sync /path/to/job.json/files`

* `chronos-nagios.py` - Nagios/Icinga style monitor of jobs
`chronos-nagios.py --hostname chronos.server.com:4400 --crit 3 --prefix etl. --prefix data.`
`chronos-nagios.py --hostname chronos.server.com:4400 --crit 3 --exclude etl.`

## Testing

``chronos-python`` uses Travis to test against multiple versions of Chronos. You can run the tests locally on any machine
with [Docker](https://www.docker.com/) and [Docker compose](https://docs.docker.com/compose/) on it.

To run the tests:

    make itests

To run against a different version of Chronos:

    CHRONOSVERSION=3.0.2 make itests


CHANGELOG.md
# Change Log

## [1.0.0](https://github.com/asher/chronos-python/tree/1.0.0) (2017-04-22)
[Full Changelog](https://github.com/asher/chronos-python/compare/v0.38.0...1.0.0)

**Closed issues:**

- Library doesn't recognize leader change [\#38](https://github.com/asher/chronos-python/issues/38)

**Merged pull requests:**

- Make a hierarchy of exceptions [\#36](https://github.com/asher/chronos-python/pull/36) ([theosotr](https://github.com/theosotr))
- Initial support for chronos 3.x for \#33 [\#35](https://github.com/asher/chronos-python/pull/35) ([thekad](https://github.com/thekad))
- Added travis deployment to pypi. Fixes \#27 [\#28](https://github.com/asher/chronos-python/pull/28) ([solarkennedy](https://github.com/solarkennedy))

## [v0.38.0](https://github.com/asher/chronos-python/tree/v0.38.0) (2016-11-16)
[Full Changelog](https://github.com/asher/chronos-python/compare/v0.37.0...v0.38.0)

**Closed issues:**

- Setup travis to push packages to pypi [\#27](https://github.com/asher/chronos-python/issues/27)

**Merged pull requests:**

- Generalize auth token [\#32](https://github.com/asher/chronos-python/pull/32) ([hylandm](https://github.com/hylandm))
- epsilon is not required field of job definition [\#31](https://github.com/asher/chronos-python/pull/31) ([xkrt](https://github.com/xkrt))
- container required fields validation [\#30](https://github.com/asher/chronos-python/pull/30) ([xkrt](https://github.com/xkrt))

## [v0.37.0](https://github.com/asher/chronos-python/tree/v0.37.0) (2016-09-29)
[Full Changelog](https://github.com/asher/chronos-python/compare/v0.36.0...v0.37.0)

**Merged pull requests:**

- Add metrics endpoint [\#26](https://github.com/asher/chronos-python/pull/26) ([Rob-Johnson](https://github.com/Rob-Johnson))

## [v0.36.0](https://github.com/asher/chronos-python/tree/v0.36.0) (2016-09-13)
**Closed issues:**

- Client doesn't deal with unauthorized access correctly [\#17](https://github.com/asher/chronos-python/issues/17)
- parent job [\#16](https://github.com/asher/chronos-python/issues/16)
- new version in pypi [\#12](https://github.com/asher/chronos-python/issues/12)
- An issue with connect method [\#2](https://github.com/asher/chronos-python/issues/2)
- Can't find the package on PyPi [\#1](https://github.com/asher/chronos-python/issues/1)

**Merged pull requests:**

- retry other chronos servers on connection errors [\#24](https://github.com/asher/chronos-python/pull/24) ([Rob-Johnson](https://github.com/Rob-Johnson))
- Add method to get undocumented job graph [\#23](https://github.com/asher/chronos-python/pull/23) ([keshavdv](https://github.com/keshavdv))
- Fix typo in README [\#22](https://github.com/asher/chronos-python/pull/22) ([keshavdv](https://github.com/keshavdv))
- Python 3 support [\#21](https://github.com/asher/chronos-python/pull/21) ([navidurrahman](https://github.com/navidurrahman))
- raise an exception when unauthorized [\#19](https://github.com/asher/chronos-python/pull/19) ([Rob-Johnson](https://github.com/Rob-Johnson))
- Update docs chronos [\#18](https://github.com/asher/chronos-python/pull/18) ([Rob-Johnson](https://github.com/Rob-Johnson))
- fail if more than one of the 'one-of' parameters exist [\#15](https://github.com/asher/chronos-python/pull/15) ([Rob-Johnson](https://github.com/Rob-Johnson))
- release 0.34.0 [\#13](https://github.com/asher/chronos-python/pull/13) ([Rob-Johnson](https://github.com/Rob-Johnson))
- add /job/stat/{job\_name} and /scheduler/stats/\* endpoints [\#11](https://github.com/asher/chronos-python/pull/11) ([Rob-Johnson](https://github.com/Rob-Johnson))
- Deal with chronos jobs that have spaces in the name [\#10](https://github.com/asher/chronos-python/pull/10) ([solarkennedy](https://github.com/solarkennedy))
- Added to pypi [\#8](https://github.com/asher/chronos-python/pull/8) ([keshavdv](https://github.com/keshavdv))
- Allow a list of servers to be passed in to the client [\#7](https://github.com/asher/chronos-python/pull/7) ([keshavdv](https://github.com/keshavdv))
- Run itests in a docker container [\#6](https://github.com/asher/chronos-python/pull/6) ([keshavdv](https://github.com/keshavdv))
- Added unit test framework [\#5](https://github.com/asher/chronos-python/pull/5) ([solarkennedy](https://github.com/solarkennedy))
- Added basic itest framework with travis integration [\#4](https://github.com/asher/chronos-python/pull/4) ([solarkennedy](https://github.com/solarkennedy))
- Correctly pass in username and password in the connect method [\#3](https://github.com/asher/chronos-python/pull/3) ([solarkennedy](https://github.com/solarkennedy))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tests/test_chronos_init.py
import json
import mock
import pytest
import httplib2

import chronos


def test_connect_accepts_single_host():
    client = chronos.ChronosClient("localhost", proto="http")
    assert client.servers == ['http://localhost']


def test_connect_accepts_list_of_hosts():
    client = chronos.ChronosClient(["host1", "host2"], proto="http")
    assert client.servers == ['http://host1', 'http://host2']


def test_connect_accepts_proto():
    client = chronos.ChronosClient("localhost", proto="fake_proto")
    assert client.servers == ['fake_proto://localhost']


def test_check_accepts_json():
    client = chronos.ChronosClient("localhost")
    fake_response = mock.Mock()
    fake_response.status = 200
    fake_content = '{ "foo": "bar" }'
    actual = client._check(fake_response, fake_content)
    assert actual == json.loads(fake_content)


def test_http_codes():
    client = chronos.ChronosClient("localhost")
    fake_response = mock.Mock()
    # all status codes 2xx and 3xx are potentially valid
    valid_codes = range(200, 399)
    for status in valid_codes:
        fake_response.status = status
        fake_content = '{ "foo": "bar" }'
        actual = client._check(fake_response, fake_content)
        assert actual == json.loads(fake_content)
    # we treat 401 in a special way, so we skip it (add 400 at the beginning)
    invalid_codes = range(402, 550)
    invalid_codes.insert(0, 400)
    for status in invalid_codes:
        fake_response.status = status
        fake_content = '{ "foo": "bar" }'
        with pytest.raises(chronos.ChronosAPIError):
            actual = client._check(fake_response, fake_content)
    # let's test 401 finally
    fake_response.status = 401
    fake_content = '{ "foo": "bar" }'
    with pytest.raises(chronos.UnauthorizedError):
        actual = client._check(fake_response, fake_content)


def test_check_returns_raw_response_when_not_json():
    client = chronos.ChronosClient("localhost")
    fake_response = mock.Mock(__getitem__=mock.Mock(return_value="not-json"))
    fake_response.status = 400
    fake_content = 'foo bar baz'
    try:
        actual = client._check(fake_response, fake_content)
    except chronos.ChronosAPIError as cap:
        actual = cap.message
    # on exceptions, the content is passed on the exception's message
    assert actual == "API returned status 400, content: %s" % (fake_content, )


def test_check_does_not_log_error_when_content_type_is_not_json():
    with mock.patch('logging.getLogger', return_value=mock.Mock(error=mock.Mock())) as mock_log:
        client = chronos.ChronosClient("localhost")
        fake_response = mock.Mock(__getitem__=mock.Mock(return_value="not-json"))
        fake_response.status = 400
        fake_content = 'foo bar baz'
        try:
            client._check(fake_response, fake_content)
        except chronos.ChronosAPIError:
            pass
        assert mock_log().error.call_count == 0


def test_check_logs_error_when_invalid_json():
    with mock.patch('logging.getLogger', return_value=mock.Mock(error=mock.Mock())) as mock_log:
        client = chronos.ChronosClient("localhost")
        fake_response = mock.Mock(__getitem__=mock.Mock(return_value="application/json"))
        fake_response.status = 400
        fake_content = 'foo bar baz'
        try:
            client._check(fake_response, fake_content)
        except chronos.ChronosAPIError:
            pass
        mock_log().error.assert_called_once_with("Response not valid json: %s" % fake_content)


def test_uses_server_list():
    client = chronos.ChronosClient(["host1", "host2", "host3"], proto="http")
    good_request = (mock.Mock(status=204), '')
    bad_request = (mock.Mock(status=500), '')

    conn_mock = mock.Mock(request=mock.Mock(side_effect=[bad_request, good_request, bad_request]))
    with mock.patch('httplib2.Http', return_value=conn_mock):
        client._call('/fake_url')
        assert conn_mock.request.call_count == 2


def test_api_error_throws_exception():
    client = chronos.ChronosClient(servers="localhost")
    mock_response = mock.Mock()
    mock_response.status = 500
    mock_request = mock.Mock(return_value=(mock_response, None))
    with mock.patch.object(httplib2.Http, 'request', mock_request):
        with pytest.raises(chronos.ChronosAPIError):
            client.list()


def test_check_missing_top_level_fields():
    client = chronos.ChronosClient(servers="localhost")
    for field in chronos.ChronosJob.fields:
        without_field = {x: 'foo' for x in filter(lambda y: y != field, chronos.ChronosJob.fields)}
        with pytest.raises(chronos.MissingFieldError):
            client._check_fields(without_field)


def test_check_missing_container_fields():
    client = chronos.ChronosClient(servers="localhost")
    for field in chronos.ChronosJob.container_fields:
        container_without_field = {x: 'foo' for x in filter(lambda y: y != field, chronos.ChronosJob.container_fields)}
        job_def = {
            "container": container_without_field,
            "command": "while sleep 10; do date =u %T; done",
            "schedule": "R/2014-09-25T17:22:00Z/PT2M",
            "name": "dockerjob",
            "owner": "test",
            "disabled": False
        }
        with pytest.raises(chronos.MissingFieldError) as excinfo:
            client._check_fields(job_def)
        assert field in str(excinfo.value)


def test_check_one_of_missing():
    client = chronos.ChronosClient(servers="localhost")
    job = {field: 'foo' for field in chronos.ChronosJob.fields}
    with pytest.raises(chronos.MissingFieldError):
        client._check_fields(job)


def test_check_one_of_all():
    client = chronos.ChronosClient(servers="localhost")
    job = {field: 'foo' for field in (chronos.ChronosJob.fields + chronos.ChronosJob.one_of)}
    with pytest.raises(chronos.OneOfViolationError):
        client._check_fields(job)


def test_check_unauthorized_raises():
    client = chronos.ChronosClient(servers="localhost")
    mock_response = mock.Mock()
    mock_response.status = 401
    with pytest.raises(chronos.UnauthorizedError):
        client._check(mock_response, '{"foo": "bar"}')


@mock.patch('chronos.ChronosJob')
def test_check_one_of_ok(patch_chronos_job):
    patch_chronos_job.one_of = ['foo', 'bar']
    patch_chronos_job.fields = ['field1', 'field2']
    job = {field: 'foo' for field in chronos.ChronosJob.fields}
    client = chronos.ChronosClient(servers="localhost")
    for one_of_field in ['foo', 'bar']:
        complete = job.copy()
        complete.update({one_of_field: 'val'})
        assert client._check_fields(complete)


@mock.patch('chronos.ChronosJob')
def test_check_one_of_more_than_one(patch_chronos_job):
    patch_chronos_job.one_of = ['foo', 'bar', 'baz']
    patch_chronos_job.fields = ['field1', 'field2']
    job = {field: 'foo' for field in (chronos.ChronosJob.fields + ['foo', 'bar'])}
    client = chronos.ChronosClient(servers="localhost")
    with pytest.raises(chronos.OneOfViolationError):
        client._check_fields(job)


@mock.patch('chronos.httplib2.Http')
def test_call_retries_on_http_error(mock_http):
    mock_call = mock.Mock(side_effect=[
        httplib2.socket.error,
        httplib2.ServerNotFoundError,
        (mock.Mock(status=200), '{"foo": "bar"}')
    ])
    mock_http.return_value = mock.Mock(
            request=mock_call
    )
    client = chronos.ChronosClient(servers=['1.2.3.4', '1.2.3.5', '1.2.3.6'])
    client._call("/foo")
    mock_call.assert_any_call('http://1.2.3.4%s/foo' % client._prefix, 'GET', body=None, headers={})
    mock_call.assert_any_call('http://1.2.3.5%s/foo' % client._prefix, 'GET', body=None, headers={})
    mock_call.assert_any_call('http://1.2.3.6%s/foo' % client._prefix, 'GET', body=None, headers={})


itests/itest_utils.py
#!/usr/bin/env python
import errno
import os
import signal
import time
from functools import wraps

import requests


class TimeoutError(Exception):
    pass


def timeout(seconds=10, error_message=os.strerror(errno.ETIME)):
    def decorator(func):
        def _handle_timeout(signum, frame):
            raise TimeoutError(error_message)

        def wrapper(*args, **kwargs):
            signal.signal(signal.SIGALRM, _handle_timeout)
            signal.alarm(seconds)
            try:
                result = func(*args, **kwargs)
            finally:
                signal.alarm(0)
            return result

        return wraps(func)(wrapper)

    return decorator


@timeout(60)
def wait_for_chronos():
    """Blocks until chronos is up"""
    # we start chronos always on port 4400
    chronos_service = 'localhost:4400'
    while True:
        print 'Connecting to chronos on %s' % chronos_service
        try:
            response = requests.get('http://%s/ping' % chronos_service, timeout=2)
        except (
            requests.exceptions.ConnectionError,
            requests.exceptions.Timeout,
        ):
            time.sleep(2)
            continue
        if response.status_code == 200:
            print "Chronos is up and running!"
            break


itests/steps/chronos_steps.py
import csv
import logging
import sys
from behave import given, when, then
import time

import chronos

log = logging.getLogger('chronos')
log.addHandler(logging.StreamHandler(sys.stdout))
log.setLevel(logging.DEBUG)
LEGACY_VERSIONS = ('2.4.0',)
DEFAULT_CHRONOS_VERSION = '3.0.2'


@given('a working chronos instance')
def working_chronos(context):
    """Adds a working chronos client as context.client for the purposes of
    interacting with it in the test."""
    if not hasattr(context, 'client'):
        chronos_servers = ['127.0.0.1:4400']
        chronos_version = context.config.userdata.get('chronos_version', DEFAULT_CHRONOS_VERSION)
        if chronos_version in LEGACY_VERSIONS:
            scheduler_api_version = None
        else:
            scheduler_api_version = 'v1'
        context.client = chronos.connect(chronos_servers, scheduler_api_version=scheduler_api_version)


@when(u'we create a trivial chronos job named "{job_name}"')
def create_trivial_chronos_job(context, job_name):
    job = {
        'command': 'echo 1',
        'name': job_name,
        'owner': '',
        'disabled': False,
        'schedule': 'R0/2014-01-01T00:00:00Z/PT60M',
    }
    chronos_version = context.config.userdata.get('chronos_version', DEFAULT_CHRONOS_VERSION)
    if chronos_version in LEGACY_VERSIONS:
        job['async'] = False
    try:
        context.client.add(job)
        context.created = True
    except:
        context.created = False
    # give it a bit of time to reflect the job in ZK
    time.sleep(0.5)


@then(u'we should be able to see the job named "{job_name}" when we list jobs')
def list_chronos_jobs_has_trivial_job(context, job_name):
    jobs = context.client.list()
    job_names = [job['name'] for job in jobs]
    assert job_name in job_names


@then(u'we should not see the job named "{job_name}" when we list jobs')
def list_chronos_jobs_hasnt_trivial_job(context, job_name):
    jobs = context.client.list()
    job_names = [job['name'] for job in jobs]
    assert job_name not in job_names


@then(u'we should be able to delete the job named "{job_name}"')
def delete_job_with_spaces(context, job_name):
    context.client.delete(job_name)
    time.sleep(0.5)


@then(u'we should not be able to see the job named "{job_name}" when we list jobs')
def not_see_job_with_spaces(context, job_name):
    jobs = context.client.list()
    job_names = [job['name'] for job in jobs]
    assert 'test chronos job with spaces' not in job_names

@then(u'we should be able to see {num_jobs:d} jobs in the job graph')
def see_job_in_graph(context, num_jobs):
    jobs = csv.reader(context.client.scheduler_graph().splitlines())
    actual = sum(1 for row in jobs)
    assert actual == num_jobs


@then(u'we should be able to see timings for the job named "{job_name}" when we look at scheduler stats')
def check_job_has_timings(context, job_name):
    stats = context.client.job_stat(job_name)
    assert stats == {'histogram': {
        'median': 0.0,
        '98thPercentile': 0.0,
        '75thPercentile': 0.0,
        '95thPercentile': 0.0,
        '99thPercentile': 0.0,
        'count': 0,
        'mean': 0.0,
        },
        'taskStatHistory': []
    }


@then(u'we should be able to see percentiles for all jobs')
def check_percentiles(context):
    ninety_ninth = context.client.scheduler_stat_99th()
    ninety_eighth = context.client.scheduler_stat_98th()
    ninety_fifth = context.client.scheduler_stat_95th()
    seventy_fifth = context.client.scheduler_stat_75th()
    for percentile in ninety_ninth, ninety_eighth, ninety_fifth, seventy_fifth:
        assert percentile == [{'jobNameLabel': 'myjob', 'time': 0.0}]


@then(u'we should be able to see the median timing for all jobs')
def check_median(context):
    medians = context.client.scheduler_stat_median()
    assert medians == [{'jobNameLabel': 'myjob', 'time': 0.0}]


@then(u'we should be able to see the mean timing for all jobs')
def check_mode(context):
    modes = context.client.scheduler_stat_median()
    assert modes == [{'jobNameLabel': 'myjob', 'time': 0.0}]

@then(u'we should be able to see metrics')
def check_metrics(context):
    metrics = context.client.metrics()
    assert isinstance(metrics, dict)
    assert metrics.has_key("version")
    assert metrics.has_key("gauges")


chronos/__init__.py
#!/usr/bin/env python

# The MIT License (MIT)
#
# Copyright (c) 2014 Asher Feldman
# Derived in part from work by Chris Zacharias (chris@imgix.com)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import httplib2
import socket
import json
import logging
import warnings

# Python 3 changed the submodule for quote
try:
    from urllib import quote
except ImportError:
    from urllib.parse import quote

SCHEDULER_API_VERSIONS = ('v1',)


class ChronosError(Exception):
    pass


class ChronosAPIError(ChronosError):
    pass


class UnauthorizedError(ChronosAPIError):
    pass


class ChronosValidationError(ChronosError):
    pass


class MissingFieldError(ChronosValidationError):
    pass


class OneOfViolationError(ChronosValidationError):
    pass


class ChronosClient(object):
    _user = None
    _password = None

    def __init__(
        self, servers, proto="http", username=None,
        password=None, extra_headers=None, level='WARN',
        scheduler_api_version='v1'
    ):
        server_list = servers if isinstance(servers, list) else [servers]
        self.servers = ["%s://%s" % (proto, server) for server in server_list]
        self.extra_headers = extra_headers
        if username and password:
            self._user = username
            self._password = password
        logging.basicConfig(format='%(asctime)s %(levelname)-8s %(message)s', level=level)
        self.logger = logging.getLogger(__name__)
        if scheduler_api_version is None:
            warnings.warn("Chronos >=3.x requires scheduler_api_version set", FutureWarning)
            self._prefix = ''
        else:
            if scheduler_api_version not in SCHEDULER_API_VERSIONS:
                raise ChronosAPIError('Wrong scheduler_api_version provided')
            self._prefix = "/%s" % (scheduler_api_version,)
        self.scheduler_api_version = scheduler_api_version

    def list(self):
        """List all jobs on Chronos."""
        return self._call("/scheduler/jobs", "GET")

    def delete(self, name):
        """Delete a job by name"""
        path = "/scheduler/job/%s" % name
        return self._call(path, "DELETE")

    def delete_tasks(self, name):
        """Terminate all tasks for a running/stuck job"""
        path = "/scheduler/task/kill/%s" % name
        return self._call(path, "DELETE")

    def run(self, name):
        """Run a job by name"""
        path = "/scheduler/job/%s" % name
        return self._call(path, "PUT")

    def add(self, job_def, update=False):
        """Schedule a new job"""
        path = "/scheduler/iso8601"
        self._check_fields(job_def)
        if "parents" in job_def:
            path = "/scheduler/dependency"
        # Cool story: chronos >= 3.0 ditched PUT and only allows POST here,
        # trying to maintain backwards compat with < 3.0
        method = "POST"
        if self.scheduler_api_version is None:
            if update:
                method = "PUT"
            else:
                method = "POST"
        return self._call(path, method, json.dumps(job_def))

    def update(self, job_def):
        """Update an existing job by name"""
        return self.add(job_def, update=True)

    def job_stat(self, name):
        """ List stats for a job """
        return self._call('/scheduler/job/stat/%s' % name, "GET")

    def scheduler_graph(self):
        return self._call('/scheduler/graph/csv', 'GET')

    def scheduler_stat_99th(self):
        return self._call('/scheduler/stats/99thPercentile', 'GET')

    def scheduler_stat_98th(self):
        return self._call('/scheduler/stats/98thPercentile', 'GET')

    def scheduler_stat_95th(self):
        return self._call('/scheduler/stats/95thPercentile', 'GET')

    def scheduler_stat_75th(self):
        return self._call('/scheduler/stats/75thPercentile', 'GET')

    def scheduler_stat_median(self):
        return self._call('/scheduler/stats/median', 'GET')

    def scheduler_stat_mean(self):
        return self._call('/scheduler/stats/mean', 'GET')

    def metrics(self):
        # for some reason, /metrics is not prefixed with the version
        return self._call('/metrics', 'GET', prefix=False)

    def _call(self, url, method="GET", body=None, headers={}, prefix=True):
        hdrs = {}
        if body:
            hdrs['Content-Type'] = "application/json"
        hdrs.update(headers)
        if prefix:
            _url = '%s%s' % (self._prefix, url, )
        else:
            _url = url
        self.logger.debug("Calling: %s %s" % (method, _url))
        if body:
            self.logger.debug("Body: %s" % body)
        conn = httplib2.Http(disable_ssl_certificate_validation=True)
        if self._user and self._password:
            conn.add_credentials(self._user, self._password)
        if self.extra_headers:
            hdrs.update(self.extra_headers)

        response = None
        servers = list(self.servers)
        while servers:
            server = servers.pop(0)
            endpoint = "%s%s" % (server, quote(_url))
            try:
                resp, content = conn.request(endpoint, method, body=body, headers=hdrs)
            except (socket.error, httplib2.ServerNotFoundError) as e:
                self.logger.error('Error while calling %s: %s. Retrying', endpoint, e.message)
                continue
            try:
                response = self._check(resp, content)
                return response
            except ChronosAPIError as e:
                self.logger.error('Error while calling %s: %s', endpoint, e.message)

        raise ChronosAPIError('No remaining Chronos servers to try')

    def _check(self, resp, content):
        status = resp.status
        self.logger.debug("status: %d" % status)
        payload = None

        if status == 401:
            raise UnauthorizedError()

        if content:
            try:
                payload = json.loads(content.decode('utf-8'))
            except ValueError:
                if resp['content-type'] == "application/json":
                    self.logger.error("Response not valid json: %s" % content)
                payload = content

        if payload is None and status != 204:
            raise ChronosAPIError("Request to Chronos API failed: status: %d, response: %s" % (status, content))

        # if the status returned is not an OK status, raise an exception
        if status >= 400:
            message = "API returned status %d, content: %s" % (status, payload,)
            # newer chronos does return the full stack trace in a message field,
            # grabbing the first 160 chars from it
            if 'message' in payload:
                self.logger.debug(payload['message'])
                message = '%s (...)' % (payload['message'][:120],)
            raise ChronosAPIError(message)

        return payload

    def _check_fields(self, job):
        fields = ChronosJob.fields
        if self.scheduler_api_version is None:
            fields.extend(ChronosJob.legacy_fields)
        for k in fields:
            if k not in job:
                raise MissingFieldError("missing required field %s" % k)

        if any(field in job for field in ChronosJob.one_of):
            if len([field for field in ChronosJob.one_of if field in job]) > 1:
                raise OneOfViolationError("Job must only include 1 of %s" % ChronosJob.one_of)
        else:
            raise MissingFieldError("Job must include one of %s" % ChronosJob.one_of)

        if "container" in job:
            container = job["container"]
            for k in ChronosJob.container_fields:
                if k not in container:
                    raise MissingFieldError("missing required container field %s" % k)

        return True


class ChronosJob(object):
    fields = [
        "command",
        "name",
        "owner",
        "disabled"
    ]
    legacy_fields = [
        "async",
    ]
    one_of = ["schedule", "parents"]
    container_fields = [
        "type",
        "image"
    ]


def connect(servers, proto="http", username=None, password=None, extra_headers=None, scheduler_api_version='v1'):
    return ChronosClient(
        servers, proto=proto, username=username, password=password,
        extra_headers=extra_headers, scheduler_api_version=scheduler_api_version
    )
